server/src-lib/Hasura/Backends/MSSQL/FromIr.hs

```diff
@@ -0,0 +1,969 @@
+-- | Translate from the DML to the TSql dialect.
+
+module Hasura.Backends.MSSQL.FromIr
+  ( fromSelectRows
+  , mkSQLSelect
+  , fromRootField
+  , fromSelectAggregate
+  , Error(..)
+  , runFromIr
+  , FromIr
+  , jsonFieldName
+  ) where
+
+import           Hasura.Prelude
+
+import qualified Data.HashMap.Strict                as HM
+import qualified Data.List.NonEmpty                 as NE
+import qualified Data.Map.Strict                    as M
+import qualified Data.Text                          as T
+import qualified Database.ODBC.SQLServer            as ODBC
+
+--import           Control.Monad.Trans.State.Strict   as S
+import           Control.Monad.Validate
+import           Data.Map.Strict                    (Map)
+import           Data.Proxy
+
+
+import qualified Hasura.GraphQL.Context             as GraphQL
+import qualified Hasura.RQL.IR.BoolExp              as IR
+import qualified Hasura.RQL.IR.OrderBy              as IR
+import qualified Hasura.RQL.IR.Select               as IR
+import qualified Hasura.RQL.Types.Column            as IR
+import qualified Hasura.RQL.Types.Common            as IR
+
+import           Hasura.Backends.MSSQL.Types        as TSQL
+import           Hasura.SQL.Backend
+
+
+--------------------------------------------------------------------------------
+-- Types
+
+-- | Most of these errors should be checked for legitimacy.
+data Error
+  = FromTypeUnsupported (IR.SelectFromG 'MSSQL Expression)
```

Comment:
@abooij A lot of these errors can be avoided by correctly parameterizing the IR as per the 'trees that grow' idea. Note that the equivalent function of `mkSQLSelect` for Postgres does not throw *any* error.